Source/m_input.c

```diff
@@ -0,0 +1,335 @@
+{
+  event = ev;
+}
+
+boolean M_InputActivated(int ident)
+{
+  switch (event->type)
+  {
+    case ev_keydown:
+      return M_InputMatchKey(ident, event->data1);
+      break;
+    case ev_mouseb_down:
+      return M_InputMatchMouseB(ident, event->data1);
+      break;
+    case ev_joyb_down:
+      return M_InputMatchJoyB(ident, event->data1);
+      break;
+    default:
+      break;
+  }
+  return false;
+}
+
+boolean M_InputDeactivated(int ident)
+{
+  switch (event->type)
+  {
+    case ev_keyup:
+      return M_InputMatchKey(ident, event->data1);
+      break;
+    case ev_mouseb_up:
+      return M_InputMatchMouseB(ident, event->data1);
+      break;
+    case ev_joyb_up:
+      return M_InputMatchJoyB(ident, event->data1);
+      break;
+    default:
+      break;
+  }
+  return false;
+}
+
+static boolean InputActive(boolean *buttons, int ident, input_type_t type)
+{
+  int i;
+  input_t *p = &composite_inputs[ident];
+
+  for (i = 0; i < p->num_inputs; ++i)
+  {
+    input_value_t *v = &p->inputs[i];
+
+    if (v->type == type && buttons[v->value])
+      return true;
+  }
+  return false;
+}
+
+boolean M_InputGameKeyActive(int ident)
+{
+  return InputActive(gamekeydown, ident, input_type_key);
+}
+
+boolean M_InputGameMouseBActive(int ident)
+{
+  return InputActive(mousebuttons, ident, input_type_mouseb);
+}
+
+boolean M_InputGameJoyBActive(int ident)
+{
+  return InputActive(joybuttons, ident, input_type_joyb);
+}
+
+boolean M_InputGameActive(int ident)
+{
+  return M_InputGameKeyActive(ident) ||
+         M_InputGameMouseBActive(ident) ||
+         M_InputGameJoyBActive(ident);
+}
+
+void M_InputGameDeactivate(int ident)
+{
+  int i;
+  input_t *p = &composite_inputs[ident];
+
+  for (i = 0; i < p->num_inputs; ++i)
+  {
+    input_value_t *v = &p->inputs[i];
+
+    switch (v->type)
+    {
+      case input_type_key:
+        if (gamekeydown[v->value])
+          gamekeydown[v->value] = false;
+        break;
+      case input_type_mouseb:
+        if (mousebuttons[v->value])
+          mousebuttons[v->value] = false;
+        break;
+      case input_type_joyb:
+        if (joybuttons[v->value])
+          joybuttons[v->value] = false;
+        break;
+      default:
+        break;
+    }
+  }
+}
+
+void M_InputReset(int ident)
+{
+  input_t *p = &composite_inputs[ident];
+
+  p->num_inputs = 0;
+}
+
+void M_InputAdd(int ident, input_value_t value)
+{
+  InputAdd(ident, value.type, value.value);
+}
+
+void M_InputSet(int ident, input_value_t *inputs)
+{
+  int i;
+  input_t *p = &composite_inputs[ident];
+
+  p->num_inputs = 0;
+  for (i = 0; i < NUM_INPUTS; ++i)
+  {
+    if (inputs[i].type > input_type_null)
+    {
+      p->inputs[i].type = inputs[i].type;
+      p->inputs[i].value = inputs[i].value;
+      p->num_inputs++;
+    }
+  }
+}
+
+struct
+{
+  int key;
+  char* name;
+} key_names[] = {
+  { 0,              "NONE" },
+  { KEYD_TAB,       "TAB"  },
+  { KEYD_ENTER,     "ENTR" },
```

Comment:
We are currently saving keys with these names in config. Maybe we should make them longer? We can also add names for the gamepad and mouse buttons.

Reply:
Indeed, why restrict oneself to four chars when it's not necessary. Everything after the `KEYD_` prefix should fit.

Reply:
The menu uses the same names, but I think we can change it. The help screen breaks, but it can be rearranged or disabled.

Reply:
Hm, okay, consistency across different uses also has its value.

Reply:
On second thought, just leave the strings as they are. They are still easy to read and nobody writes config file entries by hand anyway.